modules/openapi-generator/src/main/java/org/openapitools/codegen/templating/GeneratorTemplateContentLocator.java

```diff
@@ -0,0 +1,91 @@
+package org.openapitools.codegen.templating;
+
+import org.apache.commons.lang3.StringUtils;
+import org.openapitools.codegen.CodegenConfig;
+import org.openapitools.codegen.TemplateManager;
+import org.openapitools.codegen.api.TemplatePathLocator;
+
+import java.io.File;
+
+/**
+ * Locates templates according to {@link CodegenConfig} settings.
+ */
+public class GeneratorTemplateContentLocator implements TemplatePathLocator {
+    private final CodegenConfig codegenConfig;
+
+    /**
+     * Constructs a new instance of {@link GeneratorTemplateContentLocator} for the provided {@link CodegenConfig}
+     *
+     * @param codegenConfig A generator's configuration used for determining template file location.
+     */
+    public GeneratorTemplateContentLocator(CodegenConfig codegenConfig) {
+        this.codegenConfig = codegenConfig;
+    }
+
+    private String buildLibraryFilePath(String dir, String library, String file) {
+        return dir + File.separator + "libraries" + File.separator + library + File.separator + file;
+    }
+
+    /**
+     * Determines whether an embedded file with the specified name exists.
+     *
+     * @param name The name of the file (i.e. relative to resource root)
+     *
+     * @return true if file is an embedded resource, false if it does not exist
+     */
+    public boolean embeddedTemplateExists(String name) {
+        return this.getClass().getClassLoader().getResource(TemplateManager.getCPResourcePath(name)) != null;
+    }
+
+    /**
+     * Get the template file path with template dir prepended, and use the library template if exists.
+     *
+     * Precedence:
+     * 1) (template dir)/libraries/(library)
+     * 2) (template dir)
+     * 3) (embedded template dir)/libraries/(library)
+     * 4) (embedded template dir)
+     *
+     * Where "template dir" may be user defined and "embedded template dir" are the built-in templates for the given generator.
+     *
+     * @param relativeTemplateFile Template file
+     * @return String Full template file path
+     */
+    @Override
+    public String getFullTemplatePath(String relativeTemplateFile) {
+        CodegenConfig config = this.codegenConfig;
+
+        //check the supplied template library folder for the file
+        final String library = config.getLibrary();
+        if (StringUtils.isNotEmpty(library)) {
+            //look for the file in the library subfolder of the supplied template
+            final String libTemplateFile = buildLibraryFilePath(config.templateDir(), library, relativeTemplateFile);
+            if (new File(libTemplateFile).exists() || this.getClass().getClassLoader().getResource(libTemplateFile) != null) {
```

Comment:
JAR classpath resource path requires '/' as file separator always, must not be OS dependent, e.g. '\\' under Windows won't work!
Suggestion: use `TemplateManager.getCPResourcePath(libTemplateFile)`, may extract to similar method like `embeddedTemplateExists(String)`

Reply:
@DenisKnoepfle are you suggesting this based on code quality or based on an error you've experienced?

This method falls back to embedded templates in the last two conditions, which allows for Windows paths evaluation using exactly what you've mentioned above.

The methods as written here allow us to check for the embedded file in operating systems without windows style paths without incurring the performance overhead of doing that adjustment.

Maybe this method's comment would be better to document it as this to avoid confusion?

```
     * 1) (template dir)/libraries/(library) [Windows + *nix]
     * 2) (embedded template dir)/libraries/(library) [*nix]
     * 2) (template dir) [Windows + *nix]
     * 3) (embedded template dir) [*nix]
     * 4) (embedded template dir)/libraries/(library) [Windows only]
     * 4) (embedded template dir) [Windows only]
```

And since we use File.separator in `buildLibraryFilePath`, the file check works for both Windows and non-Windows separators consistently throughout.

The alternative that you suggest would require transforming via the regular expression in `TemplateManager.getCPResourcePath` for all operating systems, and we really shouldn't have to pay that price on non-Windows.

Reply:
Sorry for the late response, I had been on holidays.
@jimschubert I'm suggesting this based on an error I've experienced under Windows using the template path feature (resolve from classpath).

> And since we use File.separator in buildLibraryFilePath, the file check works for both Windows and non-Windows separators consistently throughout.

This is what's breaking it. You _must not use_ backslashes ('\\') in a java classpath-path, it won't work, not even under Windows. Thus it must always be a forward slash ('/') when forming a classpath-path.
Is it clear now how I meant it? Did you test it under Windows trying to find a resource on a classpath that was not in the root of a jar, e.g. _some.jar/some/deep/folder/JavaSpring/pojo.mustache_? In my case it always resorted to the embedded dir instead of using my custom mustache template.

Reply:
@DenisKnoepfle thanks for the detailed response. The issue was indeed that I was testing  only with a directory structure which mirrors the `resources` directory in our project. However, I was also testing under Git Bash which I assume is not using the same file separators. I've never done Java development under windows, so I'm not entirely sure how to attach a remote debugger to find out.

If you don't mind, I'd like to tag you in the PR I'll open if you're able to test.

Reply:
Sorry, I couldn't test it sooner, but I confirmed that it's working properly now for me under Windows (v5.0.0-beta3). Thank you!